### PR TITLE
Hawkular support for the Initial Resources

### LINKED ISSUE #13873: Support Hawkular in Initial Resources
To integrate Initial Resources with Hawkular metrics backend we have to provide Hawkular specific implementation of dataSource interface. More details in [data_source.go](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/initialresources/data_source.go).

part of #13589

### PR DESCRIPTION
**Pull request by burmanm**
Adds support for using Hawkular as the datasource for initial resources plugin. Follows the configuration of Heapster sink for easier migration to Heapster in the future.

fixes #13873

cc @piosz

### CODE REVIEW

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 1]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+
+// dataSource API
+
+func (self *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
+	q := tagQuery(kind, image, exactMatch)
+
+	m := make([]metrics.Modifier, len(self.modifiers), 2+len(self.modifiers))
+	copy(m, self.modifiers)
+
+	if namespace != "" {
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/initialresources/hawkular.go]: use ```kapi.NamespaceAll``` constant

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 2]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	}
+
+	m = append(m, metrics.Filters(metrics.TagsFilter(q), metrics.BucketsFilter(1), metrics.StartTimeFilter(start), metrics.EndTimeFilter(end), metrics.PercentilesFilter([]float64{p})))
+
+	bp, err := self.client.ReadBuckets(metrics.Counter, m...)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return int64(bp[0].Percentiles[0].Value), int64(bp[0].Samples), nil
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/initialresources/hawkular.go]: is there a null pointer risk here?  would feel generally better if did a len check before accessing.

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 3]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+		return nil, err
+	}
+	return d, nil
+}
+
+// Almost equal to the Heapster initialization, try to keep them in the sync
+func (self *hawkularSource) init() error {
+	self.modifiers = make([]metrics.Modifier, 0)
+	p := metrics.Parameters{
+		Tenant: "heapster", // This data is stored by the heapster - for no-namespace hits
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/initialresources/hawkular.go]: Does this effectively mean that if I am running Kubernetes+hawkular, that I should not create a Namespace called ```heapster```?

**Comment by derekwaynecarr** [comment]
Took an initial pass, would also appreciate @piosz input.

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 4]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	useNamespace bool
+	modifiers    []metrics.Modifier
+}
+
+const (
+	containerImageTag string = "container_base_image"
+	descriptorTag     string = "descriptor_name"
+	separator         string = "/"
+
+	defaultServiceAccountFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
```
Comment (derekwaynecarr) [inline plugin/pkg/admission/initialresources/hawkular.go]: Is anyone else making this assumption?  shouldn't this be a flag?

File: plugin/pkg/admission/initialresources/data_source.go [hunk 5]
```diff
@@ -30,6 +30,7 @@ var (
 	// TODO: figure out how to better pass password here
 	password = flag.String("ir-password", "root", "Password used for connecting to InfluxDB")
 	db       = flag.String("ir-dbname", "k8s", "InfluxDB database name which contains metrics requred by InitialResources")
+	hConfig  = flag.String("ir-hawkular", "", "Hawkular configuration URL")
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/data_source.go]: How about calling this `hawkularConfig`? `hConfig` seems a bit too abbreviated.

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 6]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+)
 
-func newHawkularSource() (dataSource, error) {
-	return nil, fmt.Errorf("hawkular source not implemented")
+// Get equivalent MetricDescriptor.Name used in the Heapster
+func heapsterEqName(kind api.ResourceName) string {
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: How about renaming this to either just `heapsterName` or `heapsterEquivalentName` - let's avoid the `Eq` abbreviation.

Also, Godoc for functions should start with the function name, e.g. `heapsterEqName gets the equivalent ...`

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 7]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	case api.ResourceMemory:
+		return "memory/usage"
+	default:
+		return ""
+	}
 }
 
-func (s *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
-	return 0, 0, fmt.Errorf("gcm source not implemented")
+// Create tagFilter query
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: Please update to match Godoc conventions (I'll stop bugging about this but please review all function Godoc comments) :smile:

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 8]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	} else {
+		q[containerImageTag] = fmt.Sprintf("%s:*", strings.Split(image, ":")[0])
+	}
+
+	return q
+}
+
+// dataSource API
+
+func (self *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: Most Go code tends not to use `self`. You might want to change it to something like `h` or `hs` or `source`.

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 9]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	if v, found := opts["tenant"]; found {
+		p.Tenant = v[0]
+	}
+
+	if v, found := opts["useServiceAccount"]; found {
+		if b, _ := strconv.ParseBool(v[0]); b {
+			// If a readable service account token exists, then use it
+			if contents, err := ioutil.ReadFile(defaultServiceAccountFile); err == nil {
+				p.Token = string(contents)
+			}
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: else, log an warning/error?

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 10]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	"io/ioutil"
+	"k8s.io/kubernetes/pkg/api"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
 	"time"
 
-	"k8s.io/kubernetes/pkg/api"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: This is normally aliased as `client`

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 11]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+	kubeClientCmd "k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: No need to alias

File: plugin/pkg/admission/initialresources/hawkular_test.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	assert "github.com/stretchr/testify/require"
+)
+
+func TestTaqQuery(t *testing.T) {
+	image := "hawkular/hawkular-metrics:latest"
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular_test.go]: Make this a constant?

File: plugin/pkg/admission/initialresources/hawkular.go [hunk 13]
```diff
@@ -17,18 +17,195 @@ limitations under the License.
...
+func tagQuery(kind api.ResourceName, image string, exactMatch bool) map[string]string {
+	q := make(map[string]string)
+
+	// Add here the descriptor_tag..
+	q[descriptorTag] = heapsterEqName(kind)
+
+	if exactMatch {
+		q[containerImageTag] = image
+	} else {
+		q[containerImageTag] = fmt.Sprintf("%s:*", strings.Split(image, ":")[0])
```
Comment (ncdc) [inline plugin/pkg/admission/initialresources/hawkular.go]: Some images can be referenced by their id/digest, such as `foo/bar@sha256:abcd1234...`. Do you need to account for that here? In other words, you can have an image repository (`foo/bar`) followed by a tag (`:sometag`) or an id (`@sha256:...`).

**Comment by derekwaynecarr** [comment]
Is there a corresponding PR that discusses how to configure Hawkular with Kubernetes?

File: plugin/pkg/admission/initialresources/hawkular.go (hunk 3 above)
Comment (burmanm) [inline plugin/pkg/admission/initialresources/hawkular.go]: No, this is only the default value (which you can override using a `tenant` parameter to hawkular), which is just the same for heapster and for this. Meaning that if user forgets to set it on Heapster and here, things would still work. If you change them on both, you can use heapster namespace for other things.

File: plugin/pkg/admission/initialresources/hawkular.go (hunk 8 above)
Comment (burmanm) [inline plugin/pkg/admission/initialresources/hawkular.go]: Yeah, this leaked from Heapster where I saw it used and copied the same naming convention in there. Although I personally like this/self and equal to separate in-module methods from imports, I'll change it.

**Comment by dchen1107** [comment]
I have a concern on this PR: With this PR in, any changes to our monitoring schema in the future might cause the breakage to this integration, which makes test / rollout / upgrade much harder by adding another metric. Can we help with our metric plumbing proposal: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/metrics-plumbing.md so that we have a standard way to access stats? cc/ @vishh

**Comment by ncdc** [comment]
@dchen1107 is it fair to say that your comment extends to the other initial resources data source types - gcm and influxdb - as well?

File: plugin/pkg/admission/initialresources/hawkular.go (hunk 4 above)
Comment (burmanm) [inline plugin/pkg/admission/initialresources/hawkular.go]: Yea, I've seen this assumption elsewhere. I added serviceAccountFile param to override this path.

**Comment by vishh** [comment]
@ncdc: Yes. I have raised similar concerns in the past while introducing the other two data sources. My intention is not to derail this PR, but highlight the fact that rolling out any schema changes will be difficult with the current approach. 
Since this is an experimental feature, can we instead focus on our effort on implementing [metrics plumbing](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/metrics-plumbing.md) and work towards making this feature production quality?

**Comment by piosz** [comment]
@burmanm thanks for the change! Overall LGTM but I'm not familiarized with Hawkular at all. I assume @derekwaynecarr and @ncdc reviewed implementation details.

**Comment by piosz** [comment]
@dchen1107 @vishh
All communication with data sources will be moved to the Heapster as described in #13693 once Prediction API will be introduced. However I wouldn't expect prediction engine to use metrics API soon because we don't know yet what kind of data are required to provide reasonable prediction.

**Comment by burmanm** [comment]
Squashed the commits and rebased from master.

**Comment by derekwaynecarr** [comment]
@burmanm - trying to focus on a way to proceed for this PR.

I think its fair to advertise the feature as experimental.  As a result, user's need to know how to experiment with the feature to even make it worthwhile to integrate in the code.  Can you add an example to ```/examples/hawkular-initial-resources``` that lets a user even know how the feature can be used in practice?

**Comment by piosz** [comment]
@k8s-bot ok to test

**Comment by vishh** [comment]
ok to test

**Comment by mwringe** [comment]
Anyone know what the hold up is on this?

**Comment by derekwaynecarr** [comment]
@mwringe @burmanm - sorry about the delay on this PR, taking a pass now.

**Comment by derekwaynecarr** [comment]
This lgtm and appears consistent with what was done already for influxdb, etc.  Sorry about the delay here.

**Comment by davidopp** [comment]
Trying rebasing and then running hack/update-all.sh, then push that commit and see if the tests pass

**Comment by burmanm** [comment]
Now the integration test is failing on something like persistent_volumes_test.go, out of my hands.

**Comment by ncdc** [comment]
@burmanm when something fails in a test and it's not related to code that you changed in your PR, it typically means that it's a test flake - something that passes most of the time, but also sometimes fails. The process in this case is to find an existing issue or create a new one, and then ask the bot to retest, referencing the flake issue. Like this (I found the issue):

@k8s-bot unit test this github flake: #21188

**Comment by davidopp** [comment]
@k8s-bot e2e test this issue: #21463

**Pull request merged**